General concepts of working with Orbit, Ephem, and Phys objects
===============================================================

[DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass) objects are used throughout sbpy to curate data. The sub-classes Orbit, Ephem, Obs and Phys are all derived from the same base class ([DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass)), meaning that they can all be used in the same way. 

The core of [DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass) is an astropy [QTable](http://docs.astropy.org/en/stable/api/astropy.table.QTable.html#astropy.table.QTable) object, which is an astropy [Table](http://docs.astropy.org/en/stable/table/) that is aware of [units](http://docs.astropy.org/en/stable/units/). [DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass) provides a number of convenience functions that make it easy to create these objects and populate them with data.

For some introductional examples on how to create, access, and modify [DataClass](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass) objects, please have a look at the [documentation](https://sbpy.readthedocs.io/en/latest/sbpy/data/dataclass.html).

In the following sections, we provide some examples that are close to real use cases.

Adding a column and converting units
---------------------------

We obtain ephemerides for our target asteroid from JPL Horizons.

In [1]:
from astropy.time import Time
import astropy.units as u
from sbpy.data import Ephem

epoch1 = Time('2018-09-01 12:00', scale='utc')
epoch2 = Time('2018-09-30 12:00', scale='utc')
eph = Ephem.from_horizons('2018 RE3',
                          location='568',
                          epochs={'start': epoch1,
                                  'stop': epoch2,
                                  'step': 5*u.d},
                          skip_daylight=True)
eph.table

<QTable masked=True length=6>
targetname    H       G    solar_presence ...  PABLon   PABLat    epoch  
             mag                          ...   deg      deg             
  str10    float64 float64      str1      ... float64  float64    object 
---------- ------- ------- -------------- ... -------- -------- ---------
(2018 RE3)    27.3    0.15                ... 338.9039   5.2605 2458363.0
(2018 RE3)    27.3    0.15                ... 340.6555   6.4863 2458368.0
(2018 RE3)    27.3    0.15                ... 340.0626  13.4255 2458373.0
(2018 RE3)    27.3    0.15                ...  76.7926  43.5756 2458378.0
(2018 RE3)    27.3    0.15                ...  80.6791   -7.097 2458383.0
(2018 RE3)    27.3    0.15                ...  83.0692 -10.5497 2458388.0

Sky motion rates are provided for both RA and Dec in units of arcsec per hour:

In [2]:
print(eph['RA*cos(Dec)_rate', 'DEC_rate'])

<QTable length=6>
RA*cos(Dec)_rate  DEC_rate 
   arcsec / h    arcsec / h
    float64       float64  
---------------- ----------
        -62.6556   36.73963
        -144.843   97.00795
        -1184.37   952.1053
        -429.792   -484.058
         -39.095   -72.7217
        0.090585   -28.8419


Let's assume that we are interested in the absolute sky motion in units of arcsec per second. It would be handy to add a column to `eph` that contains this quantity. This can be easily done in the same way as for `astropy.table` [Table objects](https://docs.astropy.org/en/stable/table/). Finally, we want to print a table containing epoch, the distance to the observer, and the corresponding absolute sky motion:

In [3]:
import numpy as np
eph['abs_rate'] = np.sqrt(eph['RA*cos(Dec)_rate']**2 + eph['DEC_rate']**2).to('arcsec/second')
eph['epoch', 'delta', 'abs_rate']

<QTable length=6>
  epoch        delta             abs_rate      
                 AU             arcsec / s     
  object      float64            float64       
--------- ---------------- --------------------
2458363.0 0.07201126546179 0.020175779667146194
2458368.0 0.04189463679977  0.04842427349261143
2458373.0 0.01285453701484   0.4221159222171138
2458378.0 0.01819643837557   0.1798132842091663
2458383.0 0.04756413980326 0.022934529529610993
2458388.0 0.07789742321182 0.008011678403266262

Filtering table content
-----------------------

We have a list of 5 asteroids that we would like to observe on the night of 2018-09-12 at the Discovery Channel telescope. However, we don't want to observe these asteroids when they have an absolute sky motion rate faster than 0.1 arcsec/s and when the Moon is up.

In [4]:
import numpy as np
import astropy.units as u
from astropy.time import Time
from sbpy.data import Ephem

# target list
targets = ['2018 RR4', '2018 RE3', '2018 RC4', '2018 RQ2', '2018 RC1']

epoch1 = Time('2018-09-13 00:00', scale='utc')
epoch2 = Time('2018-09-14 00:00', scale='utc')
eph = Ephem.from_horizons(targets,
                          location='G37',
                          epochs={'start': epoch1,
                                  'stop': epoch2,
                                  'step': 10*u.minute},
                          skip_daylight=True)
print(len(eph.table))

345


A total of 345 ephemerides have been queried for the 5 asteroids. Now we apply the absolute sky motion rate (see example above) and moon filters; if the moon is up, `eph['flags']` will be set to `'m'` (see [this document](https://ssd.jpl.nasa.gov/?horizons_doc&table_quantities#table_quantities)):

In [5]:
eph = eph[np.sqrt(eph['RA*cos(Dec)_rate']**2 + eph['DEC_rate']**2).to('arcsec/second') < 0.1*u.arcsec/u.second]
eph = eph[eph['flags'] != 'm'] 

eph

<QTable masked=True length=55>
targetname    H       G    solar_presence ...  PABLon   PABLat       epoch      
             mag                          ...   deg      deg                    
  str10    float64 float64      str1      ... float64  float64       object     
---------- ------- ------- -------------- ... -------- ------- -----------------
(2018 RC1)    26.2    0.15                ... 350.8126  4.4012 2458374.666666667
(2018 RC1)    26.2    0.15                ... 350.8204  4.3984 2458374.673611111
(2018 RC1)    26.2    0.15                ... 350.8281  4.3956 2458374.680555556
(2018 RC1)    26.2    0.15                ... 350.8357  4.3928      2458374.6875
(2018 RC1)    26.2    0.15                ... 350.8434  4.3899 2458374.694444444
(2018 RC1)    26.2    0.15                ...  350.851  4.3872 2458374.701388889
(2018 RC1)    26.2    0.15                ... 350.8586  4.3844 2458374.708333333
(2018 RC1)    26.2    0.15                ... 350.8662  4.3816 2458374.715277778
(2018 RC1)    26.2    0.15                ... 350.8737  4.3788 2458374.722222222
       ...     ...     ...            ... ...      ...     ...               ...
(2018 RC1)    26.2    0.15                ... 351.1526  4.2735 2458374.979166667
(2018 RC1)    26.2    0.15                ... 351.1605  4.2704 2458374.986111111
(2018 RC1)    26.2    0.15              A ... 351.1685  4.2673 2458374.993055556
(2018 RC1)    26.2    0.15              A ... 351.1765  4.2642         2458375.0
(2018 RC1)    26.2    0.15              A ... 351.1845  4.2611 2458375.006944444
(2018 RC1)    26.2    0.15              N ... 351.1926  4.2579 2458375.013888889
(2018 RC1)    26.2    0.15              N ... 351.2007  4.2548 2458375.020833333
(2018 RC1)    26.2    0.15              N ... 351.2089  4.2516 2458375.027777778
(2018 RC1)    26.2    0.15              C ... 351.2171  4.2484 2458375.034722222
(2018 RC1)    26.2    0.15              C ... 351.2253  4.2452 2458375.041666667

As it turns out, only asteroid 2018 RC1 is observable under these conditions.

A tiny detail: note how `eph` is a `QTable` object after the filtering and not a `Ephem` object. This is because the filtering is really applied to the underlying `QTable` object. However, you can easily turn it back into an `Ephem` object by using [DataClass.from_table](https://sbpy.readthedocs.io/en/latest/api/sbpy.data.DataClass.html#sbpy.data.DataClass.from_table).

Field Name Translations and Conversions
=======================================

Parameter names are usually not unique. For instance, `diameter` may also be referred to as `d` or `diam`, etc. sbpy acknowledges this ambiuity and provides internal translation tables for parameter names. Consider the following example:

In [6]:
from sbpy.data import Phys
import astropy.units as u
data = Phys.from_dict({'d': 10*u.km})
print(data.table)

 d  
 km 
----
10.0


In this `Phys` object, the diameter is defined as `d` and can be obtained as such. The internal translations also allow for queries with similar parameter labels:

In [7]:
data['d']

<Quantity [10.] km>

In [8]:
data['diameter']

<Quantity [10.] km>

In [9]:
data['D']

<Quantity [10.] km>

The permitted labels are defined in the list `sbpy.data.core.fieldnames`. A human readable version of this list is available in [the field name document](https://sbpy.readthedocs.io/en/latest/sbpy/data/fieldnames.html).

In addition to translations, sbpy is also using able to handle obvious property conversions, e.g., between and radius. Consider the following example:

In [10]:
data['radius']

<Quantity [5.] km>

Note that the object `data` does not contain a column `radius` before this query. However, the column is added as a result of the query:

In [11]:
data.table

<QTable length=1>
   d     radius
   km      km  
float64 float64
------- -------
   10.0     5.0